src/PopIII/popiii.cpp

```diff
@@ -0,0 +1,479 @@
+			state_cc(i, j, k, HydroSystem<PopIII>::scalar0_index + nn) = mfracs[nn] * state.rho; // we use partial densities and not mass fractions
+		}
+	});
+}
+
+template <> void RadhydroSimulation<PopIII>::ErrorEst(int lev, amrex::TagBoxArray &tags, amrex::Real /*time*/, int /*ngrow*/)
+{
+	// refine on Jeans length
+	const int N_cells = 64; // inverse of the 'Jeans number' [Truelove et al. (1997)]
+	const amrex::Real G = Gconst_;
+	const amrex::Real dx = geom[lev].CellSizeArray()[0];
+
+	auto const &prob_lo = geom[lev].ProbLoArray();
+	auto const &prob_hi = geom[lev].ProbHiArray();
+
+	for (amrex::MFIter mfi(state_new_cc_[lev]); mfi.isValid(); ++mfi) {
+		const amrex::Box &box = mfi.validbox();
+		const auto state = state_new_cc_[lev].const_array(mfi);
+		const auto tag = tags.array(mfi);
+		const int nidx = HydroSystem<PopIII>::density_index;
+
+		amrex::ParallelFor(box, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+			amrex::Real const x = prob_lo[0] + (i + static_cast<amrex::Real>(0.5)) * dx;
+			amrex::Real const y = prob_lo[1] + (j + static_cast<amrex::Real>(0.5)) * dx;
+			amrex::Real const z = prob_lo[2] + (k + static_cast<amrex::Real>(0.5)) * dx;
+
+			Real const rho = state(i, j, k, nidx);
+			Real const pressure = HydroSystem<PopIII>::ComputePressure(state, i, j, k);
+			amrex::GpuArray<Real, Physics_Traits<PopIII>::numMassScalars> massScalars = RadSystem<PopIII>::ComputeMassScalars(state, i, j, k);
+
+			amrex::Real const cs = quokka::EOS<PopIII>::ComputeSoundSpeed(rho, pressure, massScalars);
+
+			const amrex::Real l_Jeans = cs * std::sqrt(M_PI / (G * rho));
+			if (l_Jeans < (N_cells * dx) && rho > 2e-20) {
+				tag(i, j, k) = amrex::TagBox::SET;
+			}
+		});
+	}
+}
+
+template <> void RadhydroSimulation<PopIII>::ComputeDerivedVar(int lev, std::string const &dname, amrex::MultiFab &mf, const int ncomp_cc_in) const
+{
+	// compute derived variables and save in 'mf'
+	if (dname == "temperature") {
+		const int ncomp = ncomp_cc_in;
+		auto const &state = state_new_cc_[lev].const_arrays();
+		auto output = mf.arrays();
+
+		amrex::ParallelFor(mf, [=] AMREX_GPU_DEVICE(int bx, int i, int j, int k) noexcept {
+			Real const rho = state[bx](i, j, k, HydroSystem<PopIII>::density_index);
+			amrex::Real const Eint = state[bx](i, j, k, HydroSystem<PopIII>::internalEnergy_index);
+
+			amrex::GpuArray<Real, Physics_Traits<PopIII>::numMassScalars> massScalars = RadSystem<PopIII>::ComputeMassScalars(state[bx], i, j, k);
+
+			output[bx](i, j, k, ncomp) = quokka::EOS<PopIII>::ComputeTgasFromEint(rho, Eint, massScalars);
+		});
+	}
+
+	if (dname == "pressure") {
+
+		const int ncomp = ncomp_cc_in;
+		auto const &state = state_new_cc_[lev].const_arrays();
+		auto output = mf.arrays();
+
+		amrex::ParallelFor(mf, [=] AMREX_GPU_DEVICE(int bx, int i, int j, int k) noexcept {
+			amrex::Real const Pgas = HydroSystem<PopIII>::ComputePressure(state[bx], i, j, k);
+			output[bx](i, j, k, ncomp) = Pgas;
+		});
+	}
+
+	if (dname == "velx") {
```

Comment:
This is fine, but it might make sense to add vely, velz.

Reply:
Yes, but since this is just meant to be a prototype for a production simulation it is fine to show an example of a derived var.